src/core/types/keys/diffie_hellman/private_key.rs

```diff
@@ -0,0 +1,195 @@
+use crate::prelude::*;
+use crypto::keys::x25519::SecretKey as X25519PrivateKey;
+use hkdf::Hkdf;
+use k256::sha2::Sha256;
+
+#[derive(derive_more::Debug)]
+#[debug("{}", self.to_hex())]
+pub struct DiffieHellmanPrivateKey(X25519PrivateKey);
+
+impl DiffieHellmanPrivateKey {
+    pub fn generate() -> Result<Self> {
+        Exactly32Bytes::generate().to_vec().try_into()
+    }
+
+    pub fn public_key(&self) -> DiffieHellmanPublicKey {
+        self.0.public_key().into()
+    }
+
+    pub fn hkdf_key_agreement(
+        &self,
+        other: &DiffieHellmanPublicKey,
+        salt: &[u8],
+        info: &[u8],
+    ) -> Result<Exactly32Bytes> {
+        let shared_secret = self.0.diffie_hellman(&other.0).to_bytes();
+        let mut symmetric_key = [0u8; 32]; // 32-byte buffer for the symmetric key
+
+        let hkdf = Hkdf::<Sha256>::new(Some(salt), &shared_secret);
+        hkdf.expand(info, &mut symmetric_key).map_err(|err| {
```

Comment:
right, so we already have this... in `pb_hkdf_sha256` which uses password as input key material (`ikm`), and here we use `shared_secret` as `ikm`. Maybe worth generalizing. And I have already verified that it is safe to unwrap, so so this method need not be returning `Result` and you can remove the `HkdfExpandFailed` (you can verify my verification if you want :) but for our usage it is not possible to crash).

Reply:
so I propose you break out the `Hkdf` into its own module, and let it have this API:

```
pub fn hkdf_sha256(ikm: impl AsRef<[u8]>, salt: impl Into<Option<&[u8]>>, info: impl Into<Option<&[u8]>> -> Exactly32Bytes
```

and reuse it in `PbHkdfSha256` and here.